### PR TITLE
358 fix unit tests for mysql

### LINKED ISSUE #358: Failing unit tests
Moodle: MOODLE_311_STABLE
Dataflows: MOODLE_35_STABLE @ [35b9c9b](https://github.com/catalyst/moodle-tool_dataflows/commit/35b9c9bfc967b017550eb73e67be60e5414fc1e7)
PHP: 7.4
Database: MySQL5.7

Seems to pass with MariaDB here - https://github.com/catalyst/moodle-tool_dataflows/runs/7522577031?check_suite_focus=true#step:3:2103

```
Go Parallel / PHPUnit Tests / tool_dataflows.tool_dataflows_sql_reader_test.test_expressions_within_a_run 
Stack Trace
tool_dataflows\tool_dataflows_sql_reader_test::test_expressions_within_a_run 
dml_read_exception: Error reading from database (You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'int) ASC                  LIMIT 3' at line 5 SELECT *                   FROM t_config_plugins                  WHERE plugin = '--phantom_plugin--'                                ORDER BY CAST(value as int) ASC                  LIMIT 3 [array ( )])  /var/www/html/lib/dml/moodle_database.php:486 /var/www/html/lib/dml/moodle_read_slave_trait.php:293 /var/www/html/lib/dml/mysqli_native_moodle_database.php:1212 /var/www/html/admin/tool/dataflows/classes/local/step/reader_sql.php:75 /var/www/html/admin/tool/dataflows/classes/local/step/reader_sql.php:65 /var/www/html/admin/tool/dataflows/classes/local/execution/flow_engine_step.php:59 /var/www/html/admin/tool/dataflows/classes/local/execution/engine.php:320 /var/www/html/admin/tool/dataflows/classes/local/execution/engine.php:275 /var/www/html/admin/tool/dataflows/tests/tool_dataflows_sql_reader_test.php:176 /var/www/html/lib/phpunit/classes/advanced_testcase.php:80 phpvfscomposer:///var/www/html/vendor/phpunit/phpunit/phpunit:97 | 0.18 sec | 3

-- | -- | --
Go Parallel / PHPUnit Tests / tool_dataflows.tool_dataflows_variables_test.test_variables_set_at_different_scopes 
Stack Trace
tool_dataflows\tool_dataflows_variables_test::test_variables_set_at_different_scopes 
dml_read_exception: Error reading from database (You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'int) ASC                  LIMIT 10' at line 5 SELECT *                   FROM t_config_plugins                  WHERE plugin = '--phantom_plugin--'                                ORDER BY CAST(value as int) ASC                  LIMIT 10 [array ( )])  /var/www/html/lib/dml/moodle_database.php:486 /var/www/html/lib/dml/moodle_read_slave_trait.php:293 /var/www/html/lib/dml/mysqli_native_moodle_database.php:1212 /var/www/html/admin/tool/dataflows/classes/local/step/reader_sql.php:75 /var/www/html/admin/tool/dataflows/classes/local/step/reader_sql.php:65 /var/www/html/admin/tool/dataflows/classes/local/execution/flow_engine_step.php:59 /var/www/html/admin/tool/dataflows/classes/local/execution/engine.php:320 /var/www/html/admin/tool/dataflows/classes/local/execution/engine.php:275 /var/www/html/admin/tool/dataflows/tests/tool_dataflows_variables_test.php:224 /var/www/html/lib/phpunit/classes/advanced_testcase.php:80 phpvfscomposer:///var/www/html/vendor/phpunit/phpunit/phpunit:97 

Go Parallel / PHPUnit Tests / tool_dataflows.tool_dataflows_workflow_state_test.test_exporting_state_before_and_after_execution 
Stack Tracetool_dataflows\tool_dataflows_workflow_state_test::test_exporting_state_before_and_after_execution 
dml_read_exception: Error reading from database (You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'int) ASC                  LIMIT 3' at line 5 SELECT *                   FROM t_config_plugins                  WHERE plugin = '--phantom_plugin--'                                ORDER BY CAST(value as int) ASC                  LIMIT 3 [array ( )])  /var/www/html/lib/dml/moodle_database.php:486 /var/www/html/lib/dml/moodle_read_slave_trait.php:293 /var/www/html/lib/dml/mysqli_native_moodle_database.php:1212 /var/www/html/admin/tool/dataflows/classes/local/step/reader_sql.php:75 /var/www/html/admin/tool/dataflows/classes/local/step/reader_sql.php:65 /var/www/html/admin/tool/dataflows/classes/local/execution/flow_engine_step.php:59 /var/www/html/admin/tool/dataflows/classes/local/execution/engine.php:320 /var/www/html/admin/tool/dataflows/classes/local/execution/engine.php:275 /var/www/html/admin/tool/dataflows/tests/tool_dataflows_workflow_state_test.php:100 /var/www/html/lib/phpunit/classes/advanced_testcase.php:80 phpvfscomposer:///var/www/html/vendor/phpunit/phpunit/phpunit:97

Go Parallel / PHPUnit Tests / tool_dataflows.tool_dataflows_sql_reader_test.test_expressions_within_a_run
tool_dataflows\tool_dataflows_sql_reader_test::test_expressions_within_a_run
dml_read_exception: Error reading from database (You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'int) ASC
                 LIMIT 3' at line 5
SELECT *
                  FROM t_config_plugins
                 WHERE plugin = '--phantom_plugin--'
                
              ORDER BY CAST(value as int) ASC
                 LIMIT 3
[array (
)])

/var/www/html/lib/dml/moodle_database.php:486
/var/www/html/lib/dml/moodle_read_slave_trait.php:293
/var/www/html/lib/dml/mysqli_native_moodle_database.php:1212
/var/www/html/admin/tool/dataflows/classes/local/step/reader_sql.php:75
/var/www/html/admin/tool/dataflows/classes/local/step/reader_sql.php:65
/var/www/html/admin/tool/dataflows/classes/local/execution/flow_engine_step.php:59
/var/www/html/admin/tool/dataflows/classes/local/execution/engine.php:320
/var/www/html/admin/tool/dataflows/classes/local/execution/engine.php:275
/var/www/html/admin/tool/dataflows/tests/tool_dataflows_sql_reader_test.php:176
/var/www/html/lib/phpunit/classes/advanced_testcase.php:80
phpvfscomposer:///var/www/html/vendor/phpunit/phpunit/phpunit:97



Go Parallel / PHPUnit Tests / tool_dataflows.tool_dataflows_variables_test.test_variables_set_at_different_scopes
tool_dataflows\tool_dataflows_variables_test::test_variables_set_at_different_scopes
dml_read_exception: Error reading from database (You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'int) ASC
                 LIMIT 10' at line 5
SELECT *
                  FROM t_config_plugins
                 WHERE plugin = '--phantom_plugin--'
                
              ORDER BY CAST(value as int) ASC
                 LIMIT 10
[array (
)])

/var/www/html/lib/dml/moodle_database.php:486
/var/www/html/lib/dml/moodle_read_slave_trait.php:293
/var/www/html/lib/dml/mysqli_native_moodle_database.php:1212
/var/www/html/admin/tool/dataflows/classes/local/step/reader_sql.php:75
/var/www/html/admin/tool/dataflows/classes/local/step/reader_sql.php:65
/var/www/html/admin/tool/dataflows/classes/local/execution/flow_engine_step.php:59
/var/www/html/admin/tool/dataflows/classes/local/execution/engine.php:320
/var/www/html/admin/tool/dataflows/classes/local/execution/engine.php:275
/var/www/html/admin/tool/dataflows/tests/tool_dataflows_variables_test.php:224
/var/www/html/lib/phpunit/classes/advanced_testcase.php:80
phpvfscomposer:///var/www/html/vendor/phpunit/phpunit/phpunit:97

Go Parallel / PHPUnit Tests / tool_dataflows.tool_dataflows_workflow_state_test.test_exporting_state_before_and_after_execution
tool_dataflows\tool_dataflows_workflow_state_test::test_exporting_state_before_and_after_execution
dml_read_exception: Error reading from database (You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'int) ASC
                 LIMIT 3' at line 5
SELECT *
                  FROM t_config_plugins
                 WHERE plugin = '--phantom_plugin--'
                
              ORDER BY CAST(value as int) ASC
                 LIMIT 3
[array (
)])

/var/www/html/lib/dml/moodle_database.php:486
/var/www/html/lib/dml/moodle_read_slave_trait.php:293
/var/www/html/lib/dml/mysqli_native_moodle_database.php:1212
/var/www/html/admin/tool/dataflows/classes/local/step/reader_sql.php:75
/var/www/html/admin/tool/dataflows/classes/local/step/reader_sql.php:65
/var/www/html/admin/tool/dataflows/classes/local/execution/flow_engine_step.php:59
/var/www/html/admin/tool/dataflows/classes/local/execution/engine.php:320
/var/www/html/admin/tool/dataflows/classes/local/execution/engine.php:275
/var/www/html/admin/tool/dataflows/tests/tool_dataflows_workflow_state_test.php:100
/var/www/html/lib/phpunit/classes/advanced_testcase.php:80
phpvfscomposer:///var/www/html/vendor/phpunit/phpunit/phpunit:97
```

### LINKED ISSUE #364: phpunit test fails when local_aws not installed
When doing unit tests but the local_aws package is not installed.

```
Moodle 3.11.8+ (Build: 20220722), 8de4b14be7d002bf4f2491d45d731026d4e960af
Php: 7.4.28, mysqli: 5.7.34, OS: Linux 5.15.0-40-generic x86_64
PHPUnit 9.5.13 by Sebastian Bergmann and contributors.

...................................PHP Warning:  Uncaught require_once(/var/www/m311/local/aws/sdk/aws-autoloader.php): failed to open stream: No such file or directory

/var/www/m311/admin/tool/dataflows/classes/local/step/connector_s3.php:91
/var/www/m311/admin/tool/dataflows/classes/local/step/connector_s3.php:91
/var/www/m311/admin/tool/dataflows/classes/local/execution/connector_engine_step.php:47
/var/www/m311/admin/tool/dataflows/classes/local/execution/engine.php:320
/var/www/m311/admin/tool/dataflows/classes/local/execution/engine.php:275
/var/www/m311/admin/tool/dataflows/tests/tool_dataflows_secret_service_test.php:162
/var/www/m311/lib/phpunit/classes/advanced_testcase.php:80

  thrown in /var/www/m311/admin/tool/dataflows/classes/local/step/connector_s3.php on line 91
PHP Fatal error:  tool_dataflows\local\step\connector_s3::execute(): Failed opening required '/var/www/m311/local/aws/sdk/aws-autoloader.php' (include_path='/var/www/m311/lib/pear:.:/usr/share/php') in /var/www/m311/admin/tool/dataflows/classes/local/step/connector_s3.php on line 91
Engine "s3 copy": status: new
Engine "s3 copy", step "s3copy": status: new
Engine "s3 copy", step "s3copy": status: initialised
Engine "s3 copy": status: initialised, config: {"isdryrun":false}
Engine "s3 copy": status: processing

Warning: Uncaught require_once(/var/www/m311/local/aws/sdk/aws-autoloader.php): failed to open stream: No such file or directory

/var/www/m311/admin/tool/dataflows/classes/local/step/connector_s3.php:91
/var/www/m311/admin/tool/dataflows/classes/local/step/connector_s3.php:91
/var/www/m311/admin/tool/dataflows/classes/local/execution/connector_engine_step.php:47
/var/www/m311/admin/tool/dataflows/classes/local/execution/engine.php:320
/var/www/m311/admin/tool/dataflows/classes/local/execution/engine.php:275
/var/www/m311/admin/tool/dataflows/tests/tool_dataflows_secret_service_test.php:162
/var/www/m311/lib/phpunit/classes/advanced_testcase.php:80

  thrown in /var/www/m311/admin/tool/dataflows/classes/local/step/connector_s3.php on line 91

Fatal error: tool_dataflows\local\step\connector_s3::execute(): Failed opening required '/var/www/m311/local/aws/sdk/aws-autoloader.php' (include_path='/var/www/m311/lib/pear:.:/usr/share/php') in /var/www/m311/admin/tool/dataflows/classes/local/step/connector_s3.php on line 91

```

### PR DESCRIPTION
**Pull request by keevan**
- fix: unit tests for mysql, casted types in tests did not exist for mysql
- fix: issue where the s3 sdk was trying to be loaded but wasn't installed

Resolves #364 
Resolves #358

### CODE REVIEW

**Comment by keevan** [comment]
Was quite annoying but there is no cross DB type (mysql, mariadb, postgres) that can be used to use cast to an int-like value to be used for comparisons in the tests.

References used:
- https://mariadb.com/kb/en/cast/
- https://www.postgresqltutorial.com/postgresql-tutorial/postgresql-cast/
- https://dev.mysql.com/doc/refman/8.0/en/cast-functions.html#function_cast

Ended up doing a db family check instead.

File: tests/tool_dataflows_sql_reader_test.php [hunk 1]
```diff
@@ -137,13 +137,17 @@ public function test_expressions_within_a_run() {
...
+        $integertype = 'int';
+        if ($DB->get_dbfamily() === 'mysql') {
+            $integertype = 'signed';
+        }
         $sql = 'SELECT *
                   FROM {config_plugins}
                  WHERE plugin = \'' . $template['plugin'] . '\'
                 [[
-                    AND CAST(value as int) > ${{countervalue}}
+                    AND CAST(value as ' . $integertype . ') > ${{countervalue}}
```
Comment (brendanheywood) [inline tests/tool_dataflows_sql_reader_test.php:148]: shoudl this use $DB->sql_cast_** ?

Comment (keevan) [reply tests/tool_dataflows_sql_reader_test.php:148]: TIL that there's an <s>app</s> function for that, `sql_cast_char2int`. I'll give this a shot. Thank you @brendanheywood

Comment (keevan) [reply tests/tool_dataflows_sql_reader_test.php]: Much better now.

**Pull request merged**